### PR TITLE
Add LAPACK overloads for all variants of solve in Numba backend 

### LINKED ISSUE #422: BUG: `pt.linalg.solve` returns incorrect results when `mode = "NUMBA"`
### Describe the issue:

This is identical to  #382. When `assume_a != 'gen'` in `pt.linalg.solve`, the `solve_triangular` function is incorrectly invoked, and the resulting output is not correct.

### Reproducable code example:

```python
import pytensor
import pytensor.tensor as pt
import numpy as np

X = pt.dmatrix('X')
b = pt.dmatrix('b')
y1 = pt.linalg.solve(X, b, assume_a='gen')
y2 = pt.linalg.solve(X, b, assume_a='sym')

f1 = pytensor.function([X, b], y1)
f2 = pytensor.function([X, b], y2)
f1_nb = pytensor.function([X, b], y1, mode='NUMBA')
f2_nb = pytensor.function([X, b], y2, mode='NUMBA')

X_sym = np.random.normal(size=(3, 3))
X_sym = X_sym @ X_sym.T
X_inv_1 = f1(X_sym, np.eye(3))
X_inv_2 = f2(X_sym, np.eye(3))
X_inv_1_nb = f1_nb(X_sym, np.eye(3))
X_inv_2_nb = f2_nb(X_sym, np.eye(3))


# Passes, C backend, assume_a = 'gen'
np.testing.assert_allclose(X_inv_1 @ X_sym, np.eye(3), atol=1e-12)

# Passes, C backend, assume_a = 'sym'
np.testing.assert_allclose(X_inv_2 @ X_sym, np.eye(3), atol=1e-12)

# Passes, Numba backend, assume_a = 'gen'
np.testing.assert_allclose(X_inv_1_nb @ X_sym, np.eye(3), atol=1e-12)

# Fails, Numba backend, assuma_a = 'sym'
np.testing.assert_allclose(X_inv_2_nb @ X_sym, np.eye(3), atol=1e-12)
```

### Error message:

<details>

```shell
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=1e-12

Mismatched elements: 6 / 9 (66.7%)
Max absolute difference: 0.67369019
Max relative difference: 0.52399333
 x: array([[ 4.760067e-01, -2.459194e-01,  9.939378e-18],
       [-3.325732e-01,  6.636944e-01,  1.353323e-18],
       [-1.317192e-01,  6.736902e-01,  1.000000e+00]])
 y: array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])
```

</details>

### PyTensor version information:

<details>
Pytensor version 2.11.1
</details>

### Context for the issue:

_No response_

### PR DESCRIPTION
**Pull request by jessegrabowski**
## Description

Goal of this PR is to give numba mode full coverage of `scipy.linalg.solve` options. We currently only support `assume_a = "gen"`. If users select a different solver, they get incorrect results (see #422 ). This PR should fix that bug, plus add:

- [x] support for `overwrite_a` in numba mode
- [x] support for `overwrite_b` in numba mode
- [x] support for `transposed` argument (all modes)
- [ ] `lu_factor` and `lu_solve` `Ops` (all modes)
- [x] support for `assume_a = "sym"` and `assume_a = "pos"` in numba mode
- [x] support for `cho_solve` in numba mode

We get the `lu_factor` and `lu_solve` Ops kind of "for free" because I'm adding overloads for `dgetrs` and `dgetrf`. We just have to write the Ops and do the JAX dispatch. JVP for `lu_factor` is [here](https://github.com/jax-ml/jax/blob/494c15733c5b600f64c3128978c143a82537a830/jax/_src/lax/linalg.py#L1557).  Help wanted. I might decide that these Ops are out of scope for this PR and open another one.

## Related Issue

- [x] Closes #422

## Checklist

- [ ] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [x] New feature / enhancement
- [x] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pytensor--1146.org.readthedocs.build/en/1146/

### CODE REVIEW

**Comment by jessegrabowski** [comment]
This is pretty close. I just need some help with the destroy_map stuff on `Solve`. I guess this wasn't being before? The code is a bit hard to follow with all the subclassing.

File: pytensor/tensor/slinalg.py [hunk 1]
```diff
@@ -472,6 +475,7 @@ def solve_triangular(
             trans=trans,
             unit_diagonal=unit_diagonal,
             check_finite=check_finite,
+            overwrite_b=overwrite_b,
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:478]: This has to be done via inplace rewrites, and should just be ignored like we do in other OPs: https://github.com/pymc-devs/pytensor/blob/54a2c6e06b4dc7fb3b91340cf139994667120cfb/pytensor/tensor/slinalg.py#L170-L172

File: pytensor/tensor/slinalg.py [hunk 2]
```diff
@@ -581,6 +597,9 @@ def solve(
             check_finite=check_finite,
             assume_a=assume_a,
             b_ndim=b_ndim,
+            transposed=transposed,
+            overwrite_a=overwrite_a,
+            overwrite_b=overwrite_b,
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:602]: Same here

File: tests/link/numba/test_slinalg.py [hunk 3]
```diff
@@ -162,9 +170,207 @@ def test_block_diag():
...
+    A = pt.matrix("A", dtype=floatX)
+    b = b_func("b", dtype=floatX)
+
+    X = pt.linalg.solve(
+        A,
+        b,
+        lower=False,
+        assume_a=assume_a,
+        overwrite_a=overwrite_a,
+        overwrite_b=overwrite_b,
```
Comment (ricardoV94) [inline tests/link/numba/test_slinalg.py:337]: You're not testing if the inplacing is working explicitly in the numba backend. Do you want to? That would require checking the input was mutated, like we do here: https://github.com/pymc-devs/pytensor/blob/8cc489b1325995ccd375448c00b357cfb4211597/tests/tensor/test_blockwise.py#L596-L598

**Review by ricardoV94 — CHANGES_REQUESTED** [review]
What do you mean we were not doing inplace? It was already implemented for the default backend for Solve and Cholesky: https://github.com/pymc-devs/pytensor/commit/ed6ca1620526b1f6c1a84f712127c062ec6d4e07#diff-f5e942a523e3c0402aa63824184c381f5a963867125a21f119e2dab97a110d95

Just not in Numba yet. You shouldn't change anything in how the Ops are created, it's handled by the inplace rewrites. You may need to explicitly trigger them in the numba tests (the default numba mode in `compare_py_and_numba` does not include them), but I am not sure that utility actually works with inplace stuff

File: pytensor/tensor/slinalg.py (hunk 1 above)
Comment (ricardoV94) [reply pytensor/tensor/slinalg.py:478]: Instead of saying it is ignored, we should say this is applied during rewrites when it is safe to destroy the input

File: pytensor/tensor/slinalg.py [hunk 4]
```diff
@@ -407,7 +407,6 @@ def perform(self, node, inputs, outputs):
             trans=self.trans,
             unit_diagonal=self.unit_diagonal,
             check_finite=self.check_finite,
-            overwrite_b=self.overwrite_b,
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:410]: This should stay (it's the perform)

File: pytensor/tensor/slinalg.py [hunk 5]
```diff
@@ -490,14 +489,16 @@ class Solve(SolveBase):
...
     )
 
-    def __init__(self, *, assume_a="gen", **kwargs):
+    def __init__(self, *, assume_a="gen", transposed=False, **kwargs):
         if assume_a not in ("gen", "sym", "her", "pos"):
             raise ValueError(f"{assume_a} is not a recognized matrix structure")
 
         super().__init__(**kwargs)
         self.assume_a = assume_a
+        self.transposed = transposed
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:501]: Doesn't this require changes in the grads?

File: pytensor/tensor/slinalg.py (hunk 4 above)
Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:410]: whoops

File: pytensor/tensor/slinalg.py (hunk 5 above)
Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:502]: Yes, good catch

Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:518]: Something is wrong with gradients when `assume_a != 'gen'`. We agree with JAX, but not the numerical gradients for the gradient of A. According the to the numerical, it should become triangular. I think this is due to the numerical algorithm used? Not 100% sure how to proceed.

Comment (ricardoV94) [reply pytensor/tensor/slinalg.py:518]: Is this related to the transposed?

Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:518]: No, I worked out the gradients for the  transpose case. The story there is basically that we can add a rewrite for `solve(A.T, b) -> solve(A, b, transposed=True)`, and we never materialize the transpose, which can result in some savings. The only wrinkle is that in the gradients we need to do one extra transpose, because the gradient is with respect to A.T. BUT we still trade 2 transposes (one in the forward and one in the backward) for just one.

The problem here seems to be numeric. When you do `assume_a = "sym" | "pos"`, you trigger solvers that use special matrix decomposition (for example pos goes and goes cholesky solve). Numerically I think the numbers in the lower triangle of `A` won't be implicated at all, so the numerical gradient is comes back either wrong, or simplified in some way that we're not taking into account. But if that's the case, JAX doesn't take it into account either, because we agree with `jax.grad(jax.scipy.linalg.solve(A, b, assume_a).sum())`

Comment (ricardoV94) [reply pytensor/tensor/slinalg.py:518]: > and we never materialize the transpose, which can result in some savings.

Do note that transpose is always a view-only operation (no copy, just change strides). Although it may not be in some cases in Numba, #1111, which is lame / considerably a bug

Comment (ricardoV94) [reply pytensor/tensor/slinalg.py:518]: Anyway for the gradient, just roll your own version of it, if the default doesn't do the job here with these degenerate values

Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:518]: It's not clear to me if it's doing the wrong thing, or if there's something more subtle going on

Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:518]: Here's a gist showing the issue:

https://gist.github.com/jessegrabowski/4f30ebbde57758466f5b1579063ca6d0

File: pytensor/tensor/slinalg.py [hunk 6]
```diff
@@ -295,15 +295,31 @@ def L_op(self, inputs, outputs, output_gradients):
...
         c_bar = output_gradients[0]
 
-        trans_solve_op = type(self)(
-            **{
-                k: (not getattr(self, k) if k == "lower" else getattr(self, k))
-                for k in self.__props__
-            }
-        )
-        b_bar = trans_solve_op(A.T, c_bar)
+        solve_args = {k: getattr(self, k) for k in self.__props__}
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:299]: `self._props_dict()`

File: pytensor/tensor/slinalg.py [hunk 7]
```diff
@@ -295,15 +295,31 @@ def L_op(self, inputs, outputs, output_gradients):
...
+        # Some solvers can solve A.T x = b directly, without ever computing the transpose
+        has_trans = "transposed" in self.__props__
+
+        if has_trans:
+            # If the solver can do transposed solves, we do the opposite of the forward in the reverse. If we solved
+            # C = solve(A, b), then b_bar = solve(A.T, c_bar). If we solved C = solve(A.T, b), then
+            # b_bar = solve(A, c_bar)
+            solve_args["transposed"] = not solve_args["transposed"]
+            solve_op = type(self)(**solve_args)
+            b_bar = solve_op(A, c_bar)
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:310]: Do this as a rewrite?

File: pytensor/tensor/slinalg.py (hunk 5 above)
Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:685]: Reverted the transpose stuff, though I think in the end it wasn't the reason my tests were failing. I think we can revisit the transpose argument as a rewrite in a future PR.

File: pytensor/tensor/slinalg.py [hunk 8]
```diff
@@ -380,6 +381,186 @@ def cho_solve(c_and_lower, b, *, check_finite=True, b_ndim: int | None = None):
...
+            self.destroy_map = {0: [0]}
+
+    def infer_shape(self, fgraph, node, shapes):
+        n = shapes[0][0]
+        if self.permute_l:
+            return [(n, n), (n, n)]
+        elif self.p_indices:
+            return [(n,), (n, n), (n, n)]
+        else:
+            return [(n, n), (n, n), (n, n)]
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:416]: If Elemwise doesn't do it yet, we should just like Blockwise and RandomVariable infer the shape from the signature

File: pytensor/tensor/slinalg.py [hunk 9]
```diff
@@ -380,6 +381,186 @@ def cho_solve(c_and_lower, b, *, check_finite=True, b_ndim: int | None = None):
...
+        Permutation matrix, or array of integer indices for permutation matrix. Not returned if permute_l is True.
+    L: TensorVariable
+        Lower triangular matrix, or product of permutation and unit lower triangular matrices if permute_l is True.
+    U: TensorVariable
+        Upper triangular matrix
+    """
+    return cast(
+        tuple[TensorVariable, TensorVariable, TensorVariable]
+        | tuple[TensorVariable, TensorVariable],
+        LU(permute_l=permute_l, check_finite=check_finite, p_indices=p_indices)(a),
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:561]: Not Blockwise by default?

File: pytensor/link/numba/dispatch/slinalg.py [hunk 10]
```diff
@@ -266,27 +158,32 @@ def impl(A, B, trans=0, lower=False, unit_diagonal=False):
...
+    b_ndim = op.b_ndim
 
     dtype = node.inputs[0].dtype
     if str(dtype).startswith("complex"):
         raise NotImplementedError(
             "Complex inputs not currently supported by solve_triangular in Numba mode"
         )
 
-    @numba_basic.numba_njit(inline="always")
+    @numba_basic.numba_njit(inline="always", cache=True)
```
Comment (ricardoV94) [inline pytensor/link/numba/dispatch/slinalg.py:186]: We have a default handling of cache in the `numba_njit` already, why this explicit?

File: pytensor/link/numba/dispatch/slinalg.py [hunk 11]
```diff
@@ -429,3 +331,854 @@ def block_diag(*arrs):
...
+        return RCOND, int_ptr_to_val(INFO)
+
+    return impl
+
+
+def _getrf(A, overwrite_a=False) -> tuple[np.ndarray, np.ndarray, int]:
+    """
+    Placeholder for LU factorization; used by linalg.solve.
+
+    # TODO: Implement an LU_factor Op, then dispatch to this function in numba mode.
```
Comment (ricardoV94) [inline pytensor/link/numba/dispatch/slinalg.py:470]: Outdated TODO?

File: pytensor/tensor/slinalg.py (hunk 5 above)
Comment (ricardoV94) [reply pytensor/tensor/slinalg.py:685]: Last commit seems to have the transposed stuff?

File: tests/link/numba/test_slinalg.py [hunk 12]
```diff
@@ -155,16 +181,255 @@ def test_numba_Cholesky_raise_on(on_error):
...
+        b_ndim=len(b_shape),
+    )
+    f = pytensor.function(
+        [pytensor.In(A, mutable=True), pytensor.In(b, mutable=True)], X, mode="NUMBA"
+    )
+    op = f.maker.fgraph.outputs[0].owner.op
+
+    compare_numba_and_py(
+        ([A, b], [X]), inputs=[A_val.copy(), b_val.copy()], inplace=True
+    )
```
Comment (ricardoV94) [inline tests/link/numba/test_slinalg.py:379]: hmm inplace=True shouldn't mutate the original inputs, I guess it may be doing it now?

File: tests/tensor/test_slinalg.py [hunk 13]
```diff
@@ -32,6 +32,10 @@
 from tests import unittest_tools as utt
 
 
+ATOL = 1e-8 if config.floatX.endswith("64") else 1e-4
+RTOL = 1e-8 if config.floatX.endswith("64") else 1e-4
```
Comment (ricardoV94) [inline tests/tensor/test_slinalg.py:36]: I don't love these global ATOL/RTOL. Future developers may blindly use them thinking they are well thought defaults for any linalg operations?

**Comment by ricardoV94** [comment]
Missing dispatch for JAX (and possibly torch) for the LU Op?

File: pytensor/tensor/slinalg.py [hunk 14]
```diff
@@ -380,6 +382,186 @@ def cho_solve(c_and_lower, b, *, check_finite=True, b_ndim: int | None = None):
     )(A, b)
 
 
+class LU(Op):
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:385]: Missing gufunc_spec or whatever it's called, so that the Blockwise impl can avoid vectorize as scipy LU already works with batch dims?

File: pytensor/link/numba/dispatch/slinalg.py (hunk 10 above)
Comment (jessegrabowski) [reply pytensor/link/numba/dispatch/slinalg.py:186]: I was playing around with settings and forgot to delete this. Good catch.

File: pytensor/link/numba/dispatch/slinalg.py (hunk 11 above)
Comment (jessegrabowski) [reply pytensor/link/numba/dispatch/slinalg.py:470]: No, I still need lu_factor, not just lu.

File: tests/link/numba/test_slinalg.py (hunk 12 above)
Comment (jessegrabowski) [reply tests/link/numba/test_slinalg.py:379]: I might be able to remove those copies, it's being done inside `compare_numba_and_py`. I was trying a lot of different stuff to get this to work. Passing the fgraph with `pytensor.In(..., mutable=True)` inputs was breaking it.

**Review by ricardoV94 — APPROVED** [review]

**Comment by ricardoV94** [comment]
Also `test_solve_correctness` failing on float32
